perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -671,11 +671,42 @@ public void moveGroup(PerunSession sess, Group destinationGroup, Group movingGro
 				throw new GroupMoveNotAllowedException("There is already group union between moving group: " + movingGroup + " and destination group: " + destinationGroup + ".", movingGroup, destinationGroup);
 			}
 
+			//prevent existence forbidden unions between groups after moving
+			//example: there are groups "A", "A:B" and "C", where "C" is included into "A" (there is a union), if we move "C" under "A:B", existing union will change to forbidden one
+			//Get all unions for moving group and all of its subgroups
+			List<Group> movingGroupStructureUnions = new ArrayList<>();
+			movingGroupStructureUnions.addAll(getGroupUnions(sess, movingGroup, true));
+			for(Group subGroup: getAllSubGroups(sess, movingGroup)) {
+				movingGroupStructureUnions.addAll(getGroupUnions(sess, subGroup, true));
+			}
+			//remove direct relationship from moving group to it's parent group (if exists)
+			//prevent wrong exception in situation like "A", "A:B", "A:C" and moving "A:C" under "A:B", which is correct
+			if(movingGroup.getParentGroupId() != null) {
+				try {
+					Group parentGroupOfMovingGroup = getParentGroup(sess, movingGroup);
+					movingGroupStructureUnions.remove(parentGroupOfMovingGroup);
+				} catch (ParentGroupNotExistsException ex) {
+					throw new InternalErrorException("Can't find parentGroup for " + movingGroup);
+				}
+			}
+
+			//Get all group stucture of destination group (destination group all all its parent groups)
+			List<Group> destinationGroupStructure = new ArrayList<>();
```

Comment:
I think the use of Sets, with their intersection operation, would improve the performance on larger group structures.

Reply:
Changed to using intersection between sets.